lib/ansible/modules/cloud/amazon/ec2.py

```diff
@@ -1617,17 +1619,16 @@ def main():
     if not HAS_BOTO:
         module.fail_json(msg='boto required for this module')
 
-    ec2 = ec2_connect(module)
-
-    region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module)
+    try:
+        if module.params.get('region') or not module.params.get('ec2_url'):
+            ec2 = ec2_connect(module)
+        elif module.params.get('ec2_url'):
+            _, ec2_url, aws_connect_kwargs = get_aws_connection_info(module)
+            ec2 = connect_ec2_endpoint(ec2_url, **aws_connect_kwargs)
 
-    if region:
-        try:
-            vpc = connect_to_aws(boto.vpc, region, **aws_connect_kwargs)
-        except boto.exception.NoAuthHandlerFound as e:
-            module.fail_json(msg=str(e))
-    else:
-        vpc = None
+        vpc = connect_vpc(**aws_connect_kwargs)
+    except boto.exception.NoAuthHandlerFound as e:
+        module.fail_json(msg="Failed to get connection.", exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.message))
```

Comment:
This looks more like boto3 exception handling than boto - I'm pretty sure e.message is just a string, so passing it through camel_dict_to_snake_dict seems pointless.